src/iam/certhandler/modules/pkcs11.cpp

```diff
@@ -0,0 +1,995 @@
+/*
+ * Copyright (C) 2023 Renesas Electronics Corporation.
+ * Copyright (C) 2023 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "aos/iam/modules/pkcs11.hpp"
+
+#include "aos/common/tools/fs.hpp"
+#include "aos/common/tools/os.hpp"
+#include "aos/common/uuid.hpp"
+
+#include "../log.hpp"
+
+namespace aos {
+namespace iam {
+namespace certhandler {
+
+/**
+ * A helper object for search operations containing the most valuable data for certhandler.
+ */
+struct PKCS11Module::SearchObject {
+    /**
+     * Object type.
+     */
+    Optional<pkcs11::ObjectClass> mType;
+    /**
+     * Object handle.
+     */
+    pkcs11::ObjectHandle mHandle;
+    /**
+     * Object label.
+     */
+    StaticString<pkcs11::cLabelLen> mLabel;
+    /**
+     * Key identifier for public/private key pair.
+     */
+    uuid::UUID mID;
+};
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+PKCS11Module::PKCS11Module(const String& certType, const PKCS11ModuleConfig& config)
+    : mCertType(certType)
+    , mConfig(config)
+{
+}
+
+Error PKCS11Module::Init(
+    pkcs11::PKCS11Manager& pkcs11, crypto::x509::ProviderItf& x509Provider, uuid::UUIDManagerItf& uuidManager)
+{
+    mX509Provider = &x509Provider;
+    mUUIDManager = &uuidManager;
+
+    mPKCS11 = pkcs11.OpenLibrary(mConfig.mLibrary);
+    if (!mPKCS11) {
+        return AOS_ERROR_WRAP(Error::Enum::eInvalidArgument);
+    }
+
+    Error err = Error::Enum::eNone;
+
+    Tie(cTeeClientUUIDNs, err) = mUUIDManager->StringToUUID("58AC9CA0-2086-4683-A1B8-EC4BC08E01B6");
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    err = os::GetEnv(cEnvLoginType, mTeeLoginType);
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    if (!mConfig.mUserPINPath.IsEmpty() && mTeeLoginType.IsEmpty()) {
+        return AOS_ERROR_WRAP(Error::Enum::eInvalidArgument);
+    }
+
+    mConfig.mTokenLabel = GetTokenLabel();
+
+    Tie(mSlotID, err) = GetSlotID();
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    bool isOwned = false;
+
+    Tie(isOwned, err) = IsOwned();
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    if (isOwned) {
+        err = PrintInfo(mSlotID);
+        if (!err.IsNone()) {
+            return err;
+        }
+    } else {
+        LOG_DBG() << "No owned token found";
+    }
+
+    return Error::Enum::eNone;
```

Comment:
Use `ErrorEnum`

Reply:
Still not clear why ErrorEnum is needed.

Reply:
Probably yes for Error::Enum to avoid ambiguity. But it is done in this way for all Aos enums:
```c++
using ExtendedKeyUsageEnum = ExtendedKeyUsageType::Enum;
using ExtendedKeyUsage = EnumStringer<ExtendedKeyUsageType>;
```
I don't want to make the exception for Error.

Reply:
done in a separate commit